### DISCUSSION TITLE
malware as pd patches

### DISCUSSION THREAD

**Discussion by tomara-x** [pure-data/pure-data#2265] [General]
hello pure data people! i wanted to hear your thoughts about an issue i've been discussing with some of my friends.
the lack of user confirmation on some objects can allow some people to use pd patches in very bad ways. for example:
1. delete a user's entire home directory **[DON'T TRY THIS ON YOUR COMPUTER]**

https://github.com/pure-data/pure-data/assets/86204514/5a5bff0f-cadf-47e4-98dd-b607b144608b

(i wonder how this one works on windows)

2. an attacker can create a patch that reads sensitive files on a victim's computer and send them back over the network with [netsend]
3. opening links... a world of possibilities!

we were talking about ways this can be prevented, and sadly even scanning the patch before executing it to look for any of those objects is ineffective since dynamic patching allows for anything to be constructed at runtime and messages can be obfuscated.. anyways, what do you think?

**Comment by Spacechild1** [comment]
You have the same problem with any other programming language. The good news is that compared to Python or JS, Pd has very few users, so I would be surprised if a malware author deemed it a profitable target, at least in a commercial sense. On the other hand, if the author was a sociopath with a profound hatred of Pd and its users... Either way, I can't remember that being in an issue in over 25 years. But in general, you should of course only execute code from reputable sources.

**Reply by Spacechild1** [reply]
Side note: there was a thread on the Pd mailing list where we shared innocent-looking patches that would permanently damage Pd in various amusing ways. IIRC @umlaeute won by a large margin. Unfortunately, I am not able to find it.

**Reply by tomara-x** [reply]
you have a valid point, all languages have risky sides. i wanted to point out a few things:
1. murphy's law. if it can go south, it will eventually.
2. pd has no separation between source and binary. if you wanted to examine a patch before running it, you can't. unlike a python script.
3. as you pointed out, pd has a relatively small user base, but an issue like this would become more important as it grows in popularity.
4. i think a lot of people who download patches don't know or trust the author, and they have no way of doing that. i certainly don't know who'd be a reputable source :D

**Reply by Spacechild1** [reply]
> pd has no separation between source and binary. if you wanted to examine a patch before running it, you can't. unlike a python script.

That's a good point. The only real safe way would be to run the patch in a virtual machine, which of course is not something you can expect from a casual user.

**Reply by umlaeute** [reply]
> That's a good point. 

no, it's not.

Pd's source code is hard to read for a human, but of course it is open to static code analysis, which does not require it to be executed.
it is of course possible to hide malicious code from such an analysis, but the same is true for anything written in Python.


here's my entry to the "write-malware-in-pd" challenge: https://users.iem.at/zmoelnig/moneymaker/

**Reply by tomara-x** [reply]
yes, static analysis would be ineffective. however, editing a handful of objects to make them check for confirmation from the user, wouldn't be ;)

**Reply by AlexMitchellMus** [reply]
I agree that as a 'programming language' PD has the same issues as other interpreted languages.

However, we are not really talking about PD as a language here. This is more about the IDE side of PD. Which allows users to interact with the language. 

In that way, PD application is unique to allow both programming, and execution. And a system that informs the user that their patch is 'doing stuff that potentially could be dangerous' is not outlandish for an IDE to do.

**Reply by umlaeute** [reply]
i see two basic ideas here:
- provide a read-only view of a patch (where no code is executed), allowing users to inspect code before running it (and possibly whitelisting it).
  basically something like [this](https://pastie.iem.at/view/82f714cc)
  the drawback i see here, is that it is easy to just obfuscate malicious code enough to make this approach impracticable.
- provide an interface of objects to declare themselves as "dangerous".
  but how would that work in practice? 
  the problem is mainly with externals.
   - typically, anti virus systems do not ask a binary whether it contains malware - for good reason.
     so we cannot actually trust an object to be "not dangerous", even if it says it is.
   - maintaining a whitelist of known-good objects is also bound to fail (as Pd explicitly allows overriding objects - including built-in objects!)

again, i would invite you check the moneymaker example (see link above) which of your ideas would have been an effective counter measure.

**Reply by AlexMitchellMus** [reply]
I think the idea here is not to provide 100% security. That's impossible. There will be exceptions, and I'll take your word that moneymaker is quite a cunning malware.

Given that compiled externals can be installed easily, that also presents a no-go for trying to secure the system.

So that gives us a very limited scope of security, one may argue that even some small security is better than none at all?

One solution suggested by me was to concentrate on included objects in PD-Vanilla.
It looks like there are only 4 that really pose any threat to users (either via malware, or user error):

* textfile
* file
* netsend
* netreceive

A potential solution would be to enumerate a `permissions` structure such as:
```
typedef struct _permissions
{
    unsigned int p_file_write:1;
    unsigned int p_file_read:1;
    unsigned int p_net_send:1;
    unsigned int p_net_receive:1;
} t_permissions;
```
And place it inside the `_glist` struct.

In the functions of sensitive objects, we could ask what the security setting is before executing.
Example inside textfile:
```
static void qlist_write(t_qlist *x, t_symbol *filename, t_symbol *format)
{
    if (!glist_getcanvas(x->x_textbuf.b_canvas)->gl_permissions.p_file_write) {
        post("textfile does not have write permission");
        return;
    }

    int cr = 0;
    char buf[MAXPDSTRING];
    canvas_makefilename(x->x_canvas, filename->s_name,
        buf, MAXPDSTRING);
    if (!strcmp(format->s_name, "cr"))
        cr = 1;
    else if (*format->s_name)
        pd_error(x, "qlist_read: unknown flag: %s", format->s_name);
    if (binbuf_write(x->x_binbuf, buf, "", cr))
            pd_error(x, "%s: write failed", filename->s_name);
}
```

The user would then be able to change permissions if they choose.

I accept that this solution is not foolproof, however it does create some layer of protection.

**Reply by umlaeute** [reply]
> It looks like there are only 4 

- `[array]`
- `[qlist]`
- `[soundfiler]`
- `[table]`
- `[text]`
- `[writesf~]`


> And place it inside the `_glist` struct.

hmm.
why would you do that on a per-canvas basis, rather than a global setting?

and if it is a global setting, the logic could be moved to a lower level (e.g. `sys_fopen()`)

**Reply by AlexMitchellMus** [reply]
The reason it's per canvas is because the permissions are bound to the patch. 

As PD can open multiple patches, then once one patch is open, another patch could be opened that compromises the system.

However I can see that having it global would be easier to implement, and be easier for users (much less nagging).

@umlaeute, what would be useful to me is if you agree that bypassing of sensitive functions would work for basic security, regardless if permissions are per-patch, or global. If you are aware of any more cunning exploits, I think it would be helpful to know of them, as I am not a PD security expert. :-)

Obviously we can't block all writing, as patches still need to be saved. But if all reading / writing **objects** are controlled by permissions, I don't see how any malware (or user error) could run without the user first allowing it? (Given the huge caveat that this would only work for vanilla objects). But if one is installing compiled externals all bets are off anyway.

I also think such a system would be useful for users that are not aware of how specific objects work, which would also stop accidental damage. Considering PD can obfusticate functions in multiple ways (subpatches, abstractions, auto patching) it can be very useful to be informed when PD is doing things that could potentially be damaging, either by accident or malicious intent.

For example, if I download a patch that is an wavetable synth. I would like to know if hidden inside an abstraction-auto-patching area that it is downloading wavetables from the internet. Currently, a user may never look deep enough - or in the right spot to notice this. Being informed of such access is important, even if the outcome is 'ok, allow this'. As it could have implications for how a user uses the patch, or decides to modify it if they choose.

With the system I have outlined, I believe this would also protect (if enabled, and only vanilla objects) from auto-patching.  As the object itself has it's sensitive functions rendered inactive.

Thanks for your time thinking about this.

**Reply by tomara-x** [reply]
don't forget [pdcontrol], most people have terrible security on their browsers, so an attacker with the ability to open links is pretty bad

**Comment by umlaeute** [comment]
I was not saying that static analysis would be ineffective.
I was saying that SAST only gets you so far, regardless of which language you choose to analyze (otherwise, the world would be a safe place by now)

I'm not sure what exactly you are proposing to ask confirmation from the user for (there are a number of things you could ask the user for confirmation. However: how would that impact your workflow when you have to restart your patch on-stage?)

**Reply by tomara-x** [reply]
i suppose only the operations which can be abused, (delete/write files, send data over the net, open links)
i think something like that can be a setting? so if you're experimenting with untrusteed patches and want a "sandboxed" mode, then that can be activated, but on stage you probably know what you'll be running, and so you disable it?

**Reply by umlaeute** [reply]
> write files

ah yes;  `[writesf~]`

**Reply by AlexMitchellMus** [reply]
> However: how would that impact your workflow when you have to restart your patch on-stage?

I think that if you are on-stage and don't know what your patch is doing, (deleting files, sending network requests) a user probably has larger issue? What if network isn't available on-stage? What if the user moves the patch to a new on-stage laptop, which doesn't have some files the patch generated in the background?

Two comments here are mutually exclusive: 
a) only use patches you trust
b) a user has no idea what their 'trusted patch' is doing on stage

Added to this, if a user is onstage, and want's zero chance of any permissions issues (obviously they don't know what is exactly in their patch), then they would simply 'allow all' no?

**Reply by umlaeute** [reply]
i wrote:
> I'm not sure what exactly you are proposing to ask confirmation from the user for

this is an honest statement (though of course i have various ideas what you *could* mean)

so far this discussion is rather generic, but you say
>  then they would simply 'allow all' no?

which kind of implies that you have an actual implementation in mind.
(you might have discussed this already with "some of your friends", but i'm pretty sure i couldn't make it)


> Two comments here are mutually exclusive:
> - only use patches you trust
> - a user has no idea what their 'trusted patch' is doing on stage

although i think this is quite common (with various degrees of "*no*")
